A7-backward-compatibility-management.md

```diff
@@ -0,0 +1,94 @@
+* The workaround list should be well documented in gRPC repository. The doc should include the items of the list, as well as detailed description of each workaround and its corresponding compatibility issue, so that users can identify those workarounds they want to enable
+* The list should be compiled into gRPC server and shipped to users in releases
+* An item is add to the list when a backward compatibility issue is raised and a workaround is implemented on the server
+* Updated list will be pushed to users in the next release (or next import for Google3)
+* An item should be removed from the list after 6 months (roughly 4 minor version releases)
+
+The list should be implemented and maintained by the three core languages for interoperability. Wrapping languages should use surface APIs (below) to access the table.
+
+#### APIs
+```
+GRPCAPI void enable_workaround(uint32_t id);
+```
+* This API allows user to enable a particular server workaround for incompatible clients.
+* It should be exported by core languages and wrapping languages which allow users to create gRPC server. 
+
+```
+GRPCAPI bool is_client_compatible(uint32_t id, grpc_slice user_agent);
+```
+* This API allows gRPC server to check whether a workaround should be applied for a particular client, by checking the client's `user-agent`.
+* Returns `true` if
+  1. The workaround `id` is not enabled by `enable_workaround`, or
+  2. `Version comparing function` of workaround `id` returns `true` for the particular `user-agent`.
+* It should be exported to wrapping languages as a surface API.
+* For performance, this function should be called as few times as possible, e.g. once for each client and each workaround at channel creation.
+
+## Rationale
+### User-agent field
+Since gRPC currently follows a 6-weeks release cycle, using current `user-agent` field to identify a client's version should allow users to be blocked by a backward compatibility issue for at most 6 weeks. This should be acceptable in most cases.
+
+### Version comparing function
+Instead of determining client's compatibility by a minimum version number, we decided to make it more general with `version comparing function`. This allows more flexibility in determining client's compatibility, such as including client's language into consideration.
+
+### Performance
+Current API requires a client's `user-agent` string to be checked against the version comparing functions of all items in the workaround list, involving a lot of (likely duplicated) string processing. Depends on how much performance we need, an alternative is to extract key information from `user-agent` (e.g. gRPC version, language, transport, etc.) and use them as parameter of `version comparing functions`. It is a tradeoff of flexibility to performance that we may want to consider.
```

Comment:
The proposed implementation is an absolute no in terms of performance.

That said, I don't think this RFC needs to consider the actual implementation, and instead concentrate on how this is exposed and how workarounds are to be curated.